enderio-machines/src/main/java/com/enderio/machines/common/menu/EnchanterMenu.java

```diff
@@ -0,0 +1,82 @@
+import net.minecraftforge.items.wrapper.RecipeWrapper;
+
+public class EnchanterMenu extends MachineMenu<EnchanterBlockEntity>{
+    private Level level;
+
+    public EnchanterMenu(EnchanterBlockEntity blockEntity, Inventory inventory, int pContainerId) {
+        super(blockEntity, inventory, MachineMenus.ENCHANTER.get(), pContainerId);
+        this.level = blockEntity.getLevel();
+        if (blockEntity != null) {
+            addSlot(new MachineSlot(blockEntity.getItemHandlerMaster(), 0, 16, 35));
+            addSlot(new MachineSlot(blockEntity.getItemHandlerMaster(), 1, 65, 35));
+            addSlot(new MachineSlot(blockEntity.getItemHandlerMaster(), 2, 85, 35));
+            addSlot(new MachineSlot(blockEntity.getItemHandlerMaster(), 3, 144, 35) {
+                @Override
+                public void onTake(Player pPlayer, ItemStack pStack) {
+                    Optional<IEnchanterRecipe> recipe = level.getRecipeManager().getRecipeFor(MachineRecipes.Types.ENCHANTING, new RecipeWrapper(blockEntity.getItemHandlerMaster()), level);
+                    if (recipe.isPresent() && (pPlayer.experienceLevel > recipe.get().getLevelModifier() || pPlayer.isCreative())) {
+                        int amount = recipe.get().getAmount(new RecipeWrapper(blockEntity.getItemHandlerMaster()));
+                        int lapizForLevel = recipe.get().getLapizForLevel(recipe.get().getEnchantmentLevel(blockEntity.getItemHandlerMaster().getStackInSlot(1).getCount()));
+                        pPlayer.experienceLevel -= recipe.get().getLevelCost(new RecipeWrapper(blockEntity.getItemHandlerMaster()));
+                        blockEntity.getItemHandlerMaster().getStackInSlot(0).shrink(1);
+                        blockEntity.getItemHandlerMaster().getStackInSlot(1).shrink(amount);
+                        blockEntity.getItemHandlerMaster().getStackInSlot(2).shrink(lapizForLevel);
+                    }
+                    super.onTake(pPlayer, pStack);
+                }
+                
+                @Override
+                public boolean mayPickup(Player playerIn) {
+                    Optional<IEnchanterRecipe> recipe = level.getRecipeManager().getRecipeFor(MachineRecipes.Types.ENCHANTING, new RecipeWrapper(blockEntity.getItemHandlerMaster()), level);
+                    if (recipe.isPresent() && (playerIn.experienceLevel > recipe.get().getLevelCost(new RecipeWrapper(blockEntity.getItemHandlerMaster())) || playerIn.isCreative())) {
+                        return super.mayPickup(playerIn);
+                    }
+                    return false;
+                }
+            });
+        }
+        for (int y = 0; y < 3; y++) {
+            for (int x = 0; x < 9; x++) {
+                this.addSlot(new Slot(inventory, x + y * 9 + 9, 8 + x * 18, 84 + y * 18));
+            }
+        }
+
+        for (int x = 0; x < 9; x++) {
+            this.addSlot(new Slot(inventory, x, 8 + x * 18, 142));
+        }
```

Comment:
add a helper for playerslots into EIOMenu and call it in FluidTankMenu and EnchanterMenu

Reply:
In machine or synced menu?

Reply:
doesn't really matter I think